File: api/analyze_risk.py
"""
DriveSafe Youth Initiative - Python Data Science Serverless Microservice
Vercel Python Serverless Function: analyze_risk.py

Accepts JSON trip telemetry array containing time-series sensor data:
- velocity (km/h)
- g_force_x (lateral acceleration)
- g_force_y (longitudinal acceleration)
- g_force_z (vertical acceleration)
- braking_jerk (m/s^3)
- cornering_speed (km/h)

Outputs ML Risk Classification Vector & Safety Metrics.
"""

from http.server import BaseHTTPRequestHandler
import json
import math

class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def do_POST(self):
        try:
            content_length = int(self.headers.get('Content-Length', 0))
            post_data = self.rfile.read(content_length)
            payload = json.loads(post_data.decode('utf-8')) if post_data else {}

            telemetry = payload.get('telemetry', [])
            driver_id = payload.get('driver_id', 'anonymous_youth')
            trip_duration_sec = payload.get('trip_duration_sec', max(len(telemetry), 1))

            if not telemetry or not isinstance(telemetry, list):
                # Sample fallback telemetry array if empty
                telemetry = [
                    {"velocity": 45, "g_force_x": 0.1, "g_force_y": 0.2, "g_force_z": 0.98, "braking_jerk": 0.5},
                    {"velocity": 48, "g_force_x": 0.2, "g_force_y": -0.4, "g_force_z": 1.02, "braking_jerk": 1.2},
                    {"velocity": 52, "g_force_x": 0.15, "g_force_y": 0.1, "g_force_z": 1.0, "braking_jerk": 0.3},
                    {"velocity": 30, "g_force_x": 0.6, "g_force_y": -1.2, "g_force_z": 1.1, "braking_jerk": 3.8}
                ]

            # 1. Extract Statistical Metrics
            velocities = [float(point.get('velocity', 0)) for point in telemetry]
            gx = [float(point.get('g_force_x', 0)) for point in telemetry]
            gy = [float(point.get('g_force_y', 0)) for point in telemetry]
            gz = [float(point.get('g_force_z', 1.0)) for point in telemetry]
            jerks = [float(point.get('braking_jerk', 0)) for point in telemetry]

            n = len(telemetry)
            avg_velocity = sum(velocities) / n
            max_velocity = max(velocities)

            # Velocity Variance
            vel_variance = sum((v - avg_velocity) ** 2 for v in velocities) / n
            vel_std_dev = math.sqrt(vel_variance)

            # Combined Vector G-Force Magnitude
            g_magnitudes = [
                math.sqrt(x**2 + y**2 + (z - 1.0)**2) for x, y, z in zip(gx, gy, gz)
            ]
            avg_g_mag = sum(g_magnitudes) / n
            max_g_mag = max(g_magnitudes)
            g_mag_variance = sum((g - avg_g_mag) ** 2 for g in g_magnitudes) / n
            g_std_dev = math.sqrt(g_mag_variance)

            # Jerk Metrics (Braking & Sudden Acceleration)
            max_jerk = max(jerks)
            avg_jerk = sum(jerks) / n
            high_jerk_events = sum(1 for j in jerks if j > 2.5)
            harsh_cornering_events = sum(1 for x in gx if abs(x) > 0.45)
            harsh_braking_events = sum(1 for y in gy if y < -0.4)

            # 2. Machine Learning Risk Classification Heuristic Model
            # Formula: Risk Index (0 - 100)
            base_risk = 10.0
            speed_risk = min(40.0, (vel_std_dev * 1.5) + (max(0, max_velocity - 100) * 0.8))
            g_force_risk = min(30.0, (g_std_dev * 35.0) + (max_g_mag * 12.0))
            jerk_risk = min(30.0, (high_jerk_events * 5.0) + (harsh_braking_events * 4.0) + (harsh_cornering_events * 3.0))

            total_risk_score = round(min(100.0, base_risk + speed_risk + g_force_risk + jerk_risk), 1)
            safety_score = round(max(0.0, 100.0 - total_risk_score), 1)

            # Classification Category Vector
            if total_risk_score < 30.0:
                risk_category = "SAFE"
                safety_rating = "EXCELLENT"
                color_code = "#10b981" # Neon Mint
            elif total_risk_score < 60.0:
                risk_category = "MODERATE"
                safety_rating = "CAUTION_REQUIRED"
                color_code = "#f59e0b" # Amber
            else:
                risk_category = "HIGH_RISK"
                safety_rating = "CRITICAL_ATTENTION"
                color_code = "#ef4444" # Coral Red

            # Key Risk Factor Identification
            risk_factors = []
            if vel_std_dev > 12.0:
                risk_factors.append("High Speed Fluctuations (Inconsistent Cruise Control)")
            if max_g_mag > 0.6:
                risk_factors.append("Extreme Lateral/Longitudinal Acceleration Detected")
            if harsh_braking_events > 0:
                risk_factors.append(f"{harsh_braking_events} Harsh Braking Incidents Recorded")
            if harsh_cornering_events > 0:
                risk_factors.append(f"{harsh_cornering_events} Aggressive Cornering Turns Recorded")
            if not risk_factors:
                risk_factors.append("Smooth driving profile maintained throughout trip.")

            # Recommendations for Youth Drivers
            coaching_tips = []
            if "Harsh Braking" in str(risk_factors):
                coaching_tips.append("Increase follow distance to 3+ seconds to anticipate stops early.")
            if "Cornering" in str(risk_factors):
                coaching_tips.append("Brake before entering turns rather than during the curve.")
            if vel_std_dev > 10.0:
                coaching_tips.append("Maintain smooth throttle pressure to improve fuel efficiency and stability.")
            if not coaching_tips:
                coaching_tips.append("Outstanding driver focus! You qualify for safe-driver insurance discount points.")

            response_data = {
                "status": "success",
                "driver_id": driver_id,
                "trip_summary": {
                    "data_points": n,
                    "avg_velocity_kmh": round(avg_velocity, 1),
                    "max_velocity_kmh": round(max_velocity, 1),
                    "velocity_std_dev": round(vel_std_dev, 2),
                    "max_g_force": round(max_g_mag, 2),
                    "g_force_std_dev": round(g_std_dev, 3),
                    "harsh_braking_count": harsh_braking_events,
                    "harsh_cornering_count": harsh_cornering_events,
                    "high_jerk_events": high_jerk_events
                },
                "classification": {
                    "risk_score": total_risk_score,
                    "safety_score": safety_score,
                    "risk_category": risk_category,
                    "safety_rating": safety_rating,
                    "color_code": color_code,
                    "vector": [round(speed_risk, 1), round(g_force_risk, 1), round(jerk_risk, 1)]
                },
                "key_risk_factors": risk_factors,
                "coaching_tips": coaching_tips,
                "timestamp_ms": payload.get('timestamp', 1785000000)
            }

            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps(response_data, indent=2).encode('utf-8'))

        except Exception as e:
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            error_response = {"status": "error", "message": str(e)}
            self.wfile.write(json.dumps(error_response).encode('utf-8'))

    def do_GET(self):
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        info = {
            "service": "DriveSafe Youth ML Risk Analyzer Microservice",
            "version": "2.4.0",
            "runtime": "Python 3.11 Serverless Function (Vercel)",
            "supported_endpoints": ["POST /api/analyze_risk"],
            "model_architecture": "Statistical Standard Deviation Vector + Heuristic Jerk Classification"
        }
        self.wfile.write(json.dumps(info, indent=2).encode('utf-8'))

File: api/analyze-risk.ts
export default async function handler(req: any, res: any) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }

  if (req.method === 'GET') {
    return res.status(200).json({ status: 'ok', service: 'ML Telemetry Risk Analyzer' });
  }

  if (req.method === 'POST') {
    try {
      const payload = req.body || {};
      const telemetry = payload.telemetry || [];
      const driverId = payload.driver_id || 'anonymous_youth';

      if (!Array.isArray(telemetry) || telemetry.length === 0) {
        return res.status(400).json({ status: 'error', message: 'Telemetry array required' });
      }

      const velocities = telemetry.map((t: any) => Number(t.velocity || 0));
      const gx = telemetry.map((t: any) => Number(t.g_force_x || 0));
      const gy = telemetry.map((t: any) => Number(t.g_force_y || 0));
      const gz = telemetry.map((t: any) => Number(t.g_force_z || 1.0));
      const jerks = telemetry.map((t: any) => Number(t.braking_jerk || 0));

      const n = telemetry.length;
      const avgVelocity = velocities.reduce((a: number, b: number) => a + b, 0) / n;
      const maxVelocity = Math.max(...velocities);

      // Speed Variance (Normal stop-and-go driving naturally has variance; only extreme variance at high speed is penalized)
      const velVariance = velocities.reduce((sum: number, v: number) => sum + Math.pow(v - avgVelocity, 2), 0) / n;
      const velStdDev = Math.sqrt(velVariance);

      // Net Lateral & Longitudinal G-Force Magnitudes (excluding 1G gravity)
      const gMags = gx.map((x: number, i: number) => Math.sqrt(x * x + gy[i] * gy[i]));
      const avgG = gMags.reduce((a: number, b: number) => a + b, 0) / n;
      const maxG = Math.max(...gMags, 0);
      const gStdDev = Math.sqrt(gMags.reduce((sum: number, g: number) => sum + Math.pow(g - avgG, 2), 0) / n);

      let harshBrakingCount = typeof payload.harshBrakingCount === 'number' ? payload.harshBrakingCount : 0;
      let harshCorneringCount = typeof payload.harshCorneringCount === 'number' ? payload.harshCorneringCount : 0;
      let highJerkEvents = 0;

      // Scan telemetry time-series if counts were not pre-calculated
      if (harshBrakingCount === 0 && harshCorneringCount === 0 && n > 1) {
        for (let i = 0; i < n; i++) {
          const isVerticalShock = Math.abs(gz[i] - 1.0) > 0.55;
          const hasDeceleration = i > 0 && (velocities[i - 1] - velocities[i] >= 0.5);

          // Deceleration exceeding -0.48G without pure vertical shock
          if (gy[i] < -0.48 && (!isVerticalShock || hasDeceleration)) {
            harshBrakingCount++;
          }
          // Lateral turn exceeding 0.48G
          if (Math.abs(gx[i]) > 0.48 && !isVerticalShock) {
            harshCorneringCount++;
          }
          // Jerk exceeding 4.0 m/s3
          if (jerks[i] > 4.0 && !isVerticalShock) {
            highJerkEvents++;
          }
        }
      } else {
        for (let i = 0; i < n; i++) {
          const isVerticalShock = Math.abs(gz[i] - 1.0) > 0.55;
          if (jerks[i] > 4.0 && !isVerticalShock) {
            highJerkEvents++;
          }
        }
      }

      // Fair, motivating, industry-calibrated risk deductions
      // Perfect drive starts at 0 risk points (100.0 safety score)
      let speedPenalty = 0;
      if (maxVelocity > 115) { // > ~72 mph
        speedPenalty = Math.min(20.0, (maxVelocity - 115) * 0.5);
      }

      let gForcePenalty = 0;
      if (maxG > 0.45) {
        gForcePenalty += Math.min(10.0, (maxG - 0.45) * 20.0);
      }
      if (gStdDev > 0.22) {
        gForcePenalty += Math.min(10.0, (gStdDev - 0.22) * 30.0);
      }

      const brakingDeduction = Math.min(25.0, harshBrakingCount * 4.5);
      const corneringDeduction = Math.min(20.0, harshCorneringCount * 3.5);
      const jerkDeduction = Math.min(10.0, highJerkEvents * 1.5);
      const eventPenalty = brakingDeduction + corneringDeduction + jerkDeduction;

      const totalRiskDeduction = Math.min(75.0, speedPenalty + gForcePenalty + eventPenalty);
      const totalRiskScore = Number(totalRiskDeduction.toFixed(1));
      const safetyScore = Number(Math.max(25.0, 100.0 - totalRiskDeduction).toFixed(1));

      let riskCategory = 'SAFE';
      let colorCode = '#10b981';
      if (safetyScore < 65.0) {
        riskCategory = 'HIGH_RISK';
        colorCode = '#ef4444';
      } else if (safetyScore < 85.0) {
        riskCategory = 'MODERATE';
        colorCode = '#f59e0b';
      }

      const riskFactors: string[] = [];
      if (harshBrakingCount > 0) {
        riskFactors.push(`${harshBrakingCount} Harsh Braking Event${harshBrakingCount > 1 ? 's' : ''}`);
      }
      if (harshCorneringCount > 0) {
        riskFactors.push(`${harshCorneringCount} Harsh Cornering Turn${harshCorneringCount > 1 ? 's' : ''}`);
      }
      if (maxVelocity > 115) {
        const mph = Math.round(maxVelocity * 0.621371);
        riskFactors.push(`High Maximum Speed (${mph} mph / ${Math.round(maxVelocity)} km/h)`);
      }
      if (gForcePenalty > 5.0) {
        riskFactors.push('Elevated Lateral G-Force during maneuvers');
      }

      return res.status(200).json({
        status: 'success',
        driver_id: driverId,
        trip_summary: {
          data_points: n,
          avg_velocity_kmh: Number(avgVelocity.toFixed(1)),
          max_velocity_kmh: Number(maxVelocity.toFixed(1)),
          velocity_std_dev: Number(velStdDev.toFixed(2)),
          max_g_force: Number(maxG.toFixed(2)),
          g_force_std_dev: Number(gStdDev.toFixed(3)),
          harsh_braking_count: harshBrakingCount,
          harsh_cornering_count: harshCorneringCount,
          distanceKm: typeof payload.distanceKm === 'number' ? payload.distanceKm : 0
        },
        classification: {
          risk_score: totalRiskScore,
          safety_score: safetyScore,
          risk_category: riskCategory,
          color_code: colorCode,
          vector: [Number(speedPenalty.toFixed(1)), Number(gForcePenalty.toFixed(1)), Number(eventPenalty.toFixed(1))]
        },
        key_risk_factors: riskFactors.length > 0 ? riskFactors : ['Smooth, controlled driving maintained.', 'Zero harsh events detected.']
      });
    } catch (e: any) {
      return res.status(500).json({ status: 'error', message: e.message || 'Risk calculation failed' });
    }
  }

  return res.status(405).json({ status: 'error', message: 'Method not allowed' });
}
